File: src/main/java/dragonbook/inter/expr/Constant.java
package dragonbook.inter.expr;

import dragonbook.lexer.Num;
import dragonbook.lexer.Token;
import dragonbook.lexer.Word;
import dragonbook.symbols.Type;

public class Constant extends Expr {
    public Constant(Token value, Type type) {
        super(value, type);
    }

    public Constant(int value) {
        super(new Num(value), Type.INT);
    }

    public static final Constant TRUE = new Constant(Word.TRUE, Type.BOOL);
    public static final Constant FALSE = new Constant(Word.FALSE, Type.BOOL);
}

File: src/main/kotlin/dragonbook/inter/expr/Constant.kt
package dragonbook.inter.expr

import dragonbook.lexer.Num
import dragonbook.lexer.Token
import dragonbook.lexer.Word
import dragonbook.symbols.Type

data class Constant(val value: Token, override val type: Type) : Expr {

    constructor(value: Int) : this(Num(value), Type.INT)

    override val op: Token get() = value

    override fun toString(): String = op.toString()

    companion object {
        val TRUE = Constant(Word.TRUE, Type.BOOL)
        val FALSE = Constant(Word.FALSE, Type.BOOL)
    }
}